File: 240522/사각형 채우기 2/rectangle-fill-2.java
import java.util.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    static int n = sc.nextInt();
    static final int MOD = 10_007;

    public static void main(String[] args) {
        if(n == 1) {
            System.out.println(1);
            return;
        }
        if(n == 2) {
            System.out.println(3);
            return;
        }

        int[] dp = new int[n+1];
        dp[0] = 1;
        dp[1] = 1;

        for(int i = 2; i <= n; i++) {
            dp[i] = (dp[i-1]+ dp[i-2] * 2) % MOD;
        }
        System.out.println(dp[n]);
    }
}

File: 240531/사각형 채우기 2/rectangle-fill-2.kt
import java.util.*

val sc = Scanner(System.`in`)
val n = sc.nextInt()
val MOD = 10_007

fun main() {
    if(n == 1) {
        println(1)
        return
    }
    if(n == 2) {
        println(3)
        return
    }

    val dp = IntArray(n) { 0 }

    dp[1] = 1
    dp[2] = 3

    for(i in 3 .. n) {
        dp[i] = dp[i-1] * 2 + dp[i-2]
    }

    println(dp[n])
}